File: bookstore-backend/src/main/java/com/pratik/bookstore/service/BookService.java
package com.pratik.bookstore.service;

import com.pratik.bookstore.entity.Book;
import com.pratik.bookstore.entity.Category;
import com.pratik.bookstore.exception.ResourceNotFoundException;
import com.pratik.bookstore.repository.BookRepository;
import com.pratik.bookstore.repository.CategoryRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import java.util.List;

@Service
public class BookService {

    @Autowired
    private BookRepository bookRepository;

    @Autowired
    private CategoryRepository categoryRepository;

    public List<Book> getAllBooks() {
        return bookRepository.findAll();
    }

    public Book getBookById(Long id) {
        return bookRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Book not found with id: " + id));
    }

    public List<Book> getBooksByCategory(Long categoryId) {
        return bookRepository.findByCategoryId(categoryId);
    }

    public List<Book> searchBooks(String keyword) {
        return bookRepository.searchBooks(keyword);
    }

    public Book addBook(Book book, Long categoryId) {
        if (categoryId != null) {
            Category category = categoryRepository.findById(categoryId)
                    .orElseThrow(() -> new ResourceNotFoundException("Category not found with id: " + categoryId));
            book.setCategory(category);
        }
        return bookRepository.save(book);
    }

    public Book updateBook(Long id, Book bookDetails, Long categoryId) {
        Book book = getBookById(id);
        book.setTitle(bookDetails.getTitle());
        book.setAuthor(bookDetails.getAuthor());
        book.setIsbn(bookDetails.getIsbn());
        book.setPrice(bookDetails.getPrice());
        book.setStock(bookDetails.getStock());
        book.setDescription(bookDetails.getDescription());
        book.setImageUrl(bookDetails.getImageUrl());

        if (categoryId != null) {
            Category category = categoryRepository.findById(categoryId)
                    .orElseThrow(() -> new ResourceNotFoundException("Category not found with id: " + categoryId));
            book.setCategory(category);
        }

        return bookRepository.save(book);
    }

    public void deleteBook(Long id) {
        Book book = getBookById(id);
        bookRepository.delete(book);
    }
}

File: bookstore-frontend/src/services/bookService.js
import axios from 'axios';
import authService from './authService';

const API_URL = 'http://localhost:8080/api/books';

const getAuthHeader = () => {
    const token = authService.getToken();
    return token ? { Authorization: `Bearer ${token}` } : {};
};

const getAllBooks = () => {
    return axios.get(API_URL);
};

const getBookById = (id) => {
    return axios.get(`${API_URL}/${id}`);
};

const getBooksByCategory = (categoryId) => {
    return axios.get(`${API_URL}/category/${categoryId}`);
};

const searchBooks = (keyword) => {
    return axios.get(`${API_URL}/search`, { params: { keyword } });
};

const addBook = (book, categoryId) => {
    return axios.post(`${API_URL}?categoryId=${categoryId}`, book, {
        headers: getAuthHeader()
    });
};

const updateBook = (id, book, categoryId) => {
    return axios.put(`${API_URL}/${id}?categoryId=${categoryId}`, book, {
        headers: getAuthHeader()
    });
};

const deleteBook = (id) => {
    return axios.delete(`${API_URL}/${id}`, {
        headers: getAuthHeader()
    });
};

export default { getAllBooks, getBookById, getBooksByCategory, searchBooks, addBook, updateBook, deleteBook };
